_posts/sean-has-a-crystal-ball.md

```diff
@@ -0,0 +1,110 @@
+
+## Persistent or volatile
+
+Another distinction we have to make is between persistent mechanisms and volatile ones. For example, in many cases, the main difference between a dashboard and a notification/update system is that the latter can be volatile, and the former cannot. In essence, we'd expect that we can consult the order status from the web page using our laptop when ordering some goods from a mobile device. However, this is not necessarily the case with a notification system. For example, next week, we probably don't care that yesterday at 3.16 PM, virtual machine provisioning was successful. It might even be acceptable that pending notifications get lost when refreshing the browser page or closing the app.
+
+## Architectural implications 
+
+The difference sounds trivial; however, the implementation implications are not; it's not only a matter of technical difficulties; it's much more a matter of architectural choices and impact on modeling.
+
+Adding updates to a system is "easy"; it's only a matter of knowing where to plugin the logic to understand that something worth a notification happened. In the worst case, we can hook a database trigger and apply some nasty logic to incoming insert and update requests to understand what's going on in the system and, based on those, raise an appropriate update.
+
+Adding notifications can be much harder; all that is described in ["Ehi! What's up? Feedback to users' requests in distributed systems"](https://milestone.topics.it/2021/01/12/feedback-to-users-requests-in-distributed-systems.html) needs to be done. We cannot escape that level of complexity. Notifications can be categorized as updates on steroids, a much richer user experience for slightly different scenarios. There are similarities, the most important one being that both updates and notifications are a separate thing from the primary model; they are built on the side, they can be generic and used by many different services/components in the system.
+
+## Dashboards are a different beast
+
+Let's imagine that we're designing an order management system, a straightforward one where orders can be accepted and later shipped. Accepting an order is a prolonged process; the sales service performs it, and the delivery department performs shipping. Due to the nature of sales in this business, incoming orders sit in a waiting line before being processed.
+
+Now, let's imagine that we wanted to build a dashboard on top of the described process. Since the order is introduced to when it's processed, the system behavior is like the Gargantua mentioned above. Once the order is processed, we could publish an `OrderAccepted` event and build a denormalized orders dashboard view. The problem is that this "Italian public administration" style leaves users groping in the dark. The delivery department behaves similarly; no one knows anything until the shipment is shipped.
```

Comment:
```suggestion
Now, let's imagine that we want to build a dashboard on top of the described process. Since the order is introduced to when it's processed, the system behavior is like the Gargantua black hole I mentioned earlier. Once the order is processed, we could publish an `OrderAccepted` event and build a denormalized orders dashboard view. The problem is that this "Italian public administration" style leaves users groping in the dark. The delivery department behaves similarly: no one knows anything until the shipment is shipped.
```